Mathlib/RingTheory/Ideal/Maps.lean

```diff
@@ -910,3 +910,69 @@ lemma coe_ideal_map (I : Ideal A) :
     Ideal.map f I = Ideal.map (f : A →+* B) I := rfl
 
 end AlgHom
+
+noncomputable section IsPrincipal
+
+namespace Ideal
+
+open nonZeroDivisors Submodule
+
+variable (R : Type*) [CommRing R] [IsDomain R]
+
+/-- The equivalence between `Associates R` and the principal ideals of `R` defined by sending the
+class of `x` to the principal ideal generated by `x`. -/
+noncomputable def associatesEquivIsPrincipal :
+    Associates R ≃ {I : Ideal R // IsPrincipal I} := by
+  refine Equiv.ofBijective (fun x ↦ ⟨Ideal.span {Quot.out x}, ⟨Quot.out x, rfl⟩⟩)
+    ⟨fun  _ _ h ↦ ?_, fun ⟨I, hI⟩ ↦ ?_⟩
+  · rw [Subtype.mk_eq_mk, span_singleton_eq_span_singleton] at h
+    exact Quotient.out_equiv_out.mp h
+  · obtain ⟨x, hx⟩ := hI
+    refine ⟨⟦x⟧, ?_⟩
+    rw [Subtype.mk_eq_mk, hx, submodule_span_eq, span_singleton_eq_span_singleton]
+    exact Associates.mk_quot_out x
+
+@[simp]
+theorem associatesEquivIsPrincipal_apply (x : R) :
+    associatesEquivIsPrincipal R ⟦x⟧ = Ideal.span {x} := by
+  rw [associatesEquivIsPrincipal, Equiv.ofBijective_apply, span_singleton_eq_span_singleton]
+  exact Associates.mk_quot_out x
+
+theorem associatesEquivIsPrincipal_mem_nonZeroDivisors {x : R} :
+    ↑(Ideal.associatesEquivIsPrincipal R ⟦x⟧) ∈ (Ideal R)⁰ ↔ x ∈ R⁰ := by
+  rw [Ideal.associatesEquivIsPrincipal_apply, mem_nonZeroDivisors_iff, mem_nonZeroDivisors_iff,
+    ← not_iff_not]
+  push_neg
+  refine ⟨fun ⟨I, hI₁, hI₂⟩ ↦ ?_, fun ⟨y, hy₁, hy₂⟩ ↦ ?_⟩
+  · rw [Ideal.zero_eq_bot, Submodule.ne_bot_iff] at hI₂
+    refine ⟨hI₂.choose, ?_, hI₂.choose_spec.2⟩
+    · suffices hI₂.choose * x ∈ I * Ideal.span {x} by
+        rwa [hI₁] at this
+      rw [Ideal.mem_mul_span_singleton]
+      exact ⟨hI₂.choose, hI₂.choose_spec.1, rfl⟩
+  · refine ⟨Ideal.span {y}, ?_, ?_⟩
+    · rw [Ideal.span_singleton_mul_span_singleton, hy₁, Set.singleton_zero, Ideal.span_zero,
+      Ideal.zero_eq_bot]
+    · rw [Ideal.zero_eq_bot, Submodule.ne_bot_iff]
+      exact ⟨y, Ideal.mem_span_singleton_self y, hy₂⟩
+
+/-- A version of `Ideal.associatesEquivIsPrincipal` for non-zero-divisor generators. -/
+def associatesNonZeroDivisorsEquivIsPrincipal :
+    Associates R⁰ ≃ {I : (Ideal R)⁰ // IsPrincipal (I : Ideal R)} :=
+  calc Associates R⁰ ≃ (Associates R)⁰ := associatesNonZeroDivisorsEquiv.toEquiv.symm
+    _ ≃ {I : {I : Ideal R // IsPrincipal I} // I.1 ∈ (Ideal R)⁰} :=
+      Equiv.subtypeEquiv (Ideal.associatesEquivIsPrincipal R)
+        (fun x ↦ by rw [← Associates.quot_out x, mk_mem_nonZeroDivisors_associates,
+          Ideal.associatesEquivIsPrincipal_mem_nonZeroDivisors])
+    _ ≃ {I : Ideal R // IsPrincipal I ∧ I ∈ (Ideal R)⁰} :=
+      Equiv.subtypeSubtypeEquivSubtypeInter (fun I ↦ IsPrincipal I) (fun I ↦ I ∈ (Ideal R)⁰)
+    _ ≃ {I : Ideal R // I ∈ (Ideal R)⁰ ∧ IsPrincipal I} := Equiv.setCongr (by simp_rw [and_comm])
+    _ ≃ {I : (Ideal R)⁰ // IsPrincipal I.1} := (Equiv.subtypeSubtypeEquivSubtypeInter _ _).symm
+
```

Comment:
Would you like to add a (simp) lemma saying that the effect of `associatesEquivIsPrincipal` on a non-zero-divisor is the same as the effect of `associatesNonZeroDivisorsEquivIsPrincipal`?

Reply:
Don't you mean the other way around?

Reply:
That is simplify from `associatesNonZeroDivisorsEquivIsPrincipal` to `associatesEquivIsPrincipal`?

Reply:
Yes, sure, I am sorry: I meant to "connect" them in the right way, that is the one you propose.